### PR TITLE
Next-formalizationFull.hs

### PR DESCRIPTION
**Pull request by uaksu**
Average duration for events which change the ownership of the subject is added.
To handle agreements, I updated OEMEvent.oemEvtSubject characteristic as optional using Maybe. So far I used paymentRequired as a separator between invoice and delivery. But, here I see it is set as Always for a delivery also. If we say paymentRequired=Always for a delivery then there will be no difference between it and an invoice. How shall we continue?

### CODE REVIEW

**Comment by uaksu** [comment]
Average duration for events which change the ownership of the subject is added.
To handle agreements, I updated OEMEvent.oemEvtSubject characteristic as optional using Maybe. So far I used paymentRequired as a separator between invoice and delivery. But, here I see it is set as Always for a delivery also. If we say paymentRequired=Always for a delivery then there will be no difference between it and an invoice. How shall we continue?

File: Next-formalizationFull.hs [hunk 1]
```diff
@@ -153,14 +181,29 @@ deduceStock es = [ countStock ent es | ent <- getEntities es]
...
+
+getPlannedEndDate:: EventInstance -> Date
+getPlannedEndDate e = (getPlannedDate e) !!2
+
 -- actual
 getActualDate:: EventInstance -> TimeStamps
 getActualDate e = evtInstActual e
 
+getActualStarDate::EventInstance -> Date
+getActualStarDate e =  (getActualDate e) !! 1
```
Comment (evdven) [inline Next-formalizationFull.hs]: why have implicit ordering? This is a bit tricky

File: Next-formalizationFull.hs [hunk 2]
```diff
@@ -186,10 +229,65 @@ afterClosed (d, m, y) func int evt = not (beforeOpen (d, m, y) func int evt)
 inperiod:: Date -> (EventInstance -> TimeStamps) -> Int -> Date -> (EventInstance -> TimeStamps) -> Int -> EventInstance -> Bool
 inperiod t1 func1 int1 t2 func2 int2 evt = afterClosed t1 func1 int1 evt && beforeClosed t2 func2 int2 evt
 
+-- is done
+isdone:: EventInstance -> Bool
+isdone e = if List.genericLength (getActualDate e) == 3 then True else False
```
Comment (evdven) [inline Next-formalizationFull.hs]: hmmm. can we have more gentle definition here?

File: Next-formalizationFull.hs [hunk 3]
```diff
@@ -83,20 +90,32 @@ good2 = EntityInstance{entRef = goodEntity, entInstName = "Auto", entInstType =
...
+delivery1a = EventInstance{evtRef = deliveryEvent, evtInstName = "delivery1a", evtInstStyle = Regular, evtInstHasParty = True, evtInstParty = customer1, evtInstHasSubject = True, evtInstSubject = product1, evtInstHasAmount = True, evtInstHasValue = True, evtInstDirection  = DirOut, evtInstChangeOfOwnership = Always, evtInstTypeOfOwnership = TradeItem, evtInstPaymentRequired = Never, evtInstPlanned = [(1, 1, 2017), (5, 1, 2017), (10, 2, 2017)], evtInstActual = [(3, 1, 2017), (10, 2, 2017), (16, 3, 2017)], evtInstAdministrative = [(1, 1, 2017), (5, 1, 2017), (10, 2, 2017)], evtInstAmount = 5, evtInstValue = [100, 500]}
```
Comment (evdven) [inline Next-formalizationFull.hs]: can we have a clean explicit meaning on the dates which are irrespective of their order?

**Review by evdven — CHANGES_REQUESTED** [review]

File: Next-formalizationFull.hs (hunk 3 above)
Comment (uaksu) [reply Next-formalizationFull.hs]: Based on Dennis' formalizations (https://www.dropbox.com/sh/wotluzcfcccm3rp/AAAOL7IqJk2OV96goCXHUXvPa?dl=0) the lifecycle (status) of an event instance will be determined using the values inside the plannedDate and actualDate.. For example, PlannedDate has three dates: the first one is for ToDo, the second one is for planned start, and the last one for the planned end date. However, I could not figure out why do we need ToDo date for ActualDate.  I will add explanation to the file.

File: Next-formalizationFull.hs (hunk 2 above)
Comment (uaksu) [reply Next-formalizationFull.hs]: Do you mean should I add an explanation for isdone function or something else? isdone function looks at the actualDate field of an event instance and determines whether the event instance has a Done date.

Comment (evdven) [reply Next-formalizationFull.hs]: I'm looking for a more clean declarative meaninfull definition like

type Date = (Integer, Integer, Integer)
data Timing = Timing {begin :: Date, end :: Date} deriving (Eq, Show)

data EventInstance = EventInstance {name :: String, planned :: Timing, actual :: Timing} deriving (Eq, Show)

event1 = EventInstance{name = "My first event!", planned = Timing{begin = (5, 1, 2017), end = (10, 2, 2017)}, actual = Timing {begin = (6, 1, 2017), end = (16, 3, 2017)}}

showme = begin (planned event1)

**Pull request merged**